### ISSUE TITLE
Split screen to fit 4 windows horizontally.

### ISSUE THREAD

**Issue by kamzata** [rxhanson/Rectangle#210]
I own a LG 49WL95C-W with a 5120x1440px resolution. It would be nice fit 4 windows horizontally.

**Comment by rxhanson** [comment]
I don't have plans to add this in myself, but would consider a pull request for it.

**Comment by kamzata** [comment]
> I don't have plans to add this in myself, but would consider a pull request for it.

It would be great for these new size 32:9 monitors. 3 windows fit good but fit 4 can be more productive and still usable since 1 of them would be 1280 wide.

**Comment by rxhanson** [comment]
I changed my mind - I'm going to add this in to the next release. I was going to add sixths and it makes sense to just add this in at the same time.

**Comment by kamzata** [comment]
> I changed my mind - I'm going to add this in to the next release. I was going to add sixths and it makes sense to just add this in at the same time.

This is a great news. By the way, could I ask you... what is the relation between Rectangle and Magnet?

**Comment by rxhanson** [comment]
Mind giving the beta a test? https://github.com/rxhanson/Rectangle/releases/tag/0.33

**Comment by kamzata** [comment]
> Mind giving the beta a test? https://github.com/rxhanson/Rectangle/releases/tag/0.33

Glad to be your guinea pig.

**Comment by Tinche** [comment]
Hah, excellent! I too am looking at getting a 49-inch ultrawide and am looking for tools to optimize my flow. I was going to suggest fourths but I see there's already work being done on the more-sizes branch.

Apart from first-forth and last-forth, could I suggest a command for the middle two fourths (middle half) as well? The idea being, on an ultrawide, the window layout being 25%-50%(main focus)-25%.

**Comment by rxhanson** [comment]
@kamzata You mean dragging to screen edges? What's your expected behavior here?

@Tinche That's in the beta, if you wanted to give it a try before I get a release out the door. The beta is stable, it's mainly held up because I have to prep it for localization and I had a couple more things I was going to add in.

**Comment by kamzata** [comment]
> @kamzata You mean dragging to screen edges? What's your expected behavior here?
> 
> @Tinche That's in the beta, if you wanted to give it a try before I get a release out the door. The beta is stable, it's mainly held up because I have to prep it for localization and I had a couple more things I was going to add in.

Yeah, I meant just that. My expected behavior is the same as for "3 windows" setup, but just for 4.

**Comment by rxhanson** [comment]
@kamzata Gotcha. That makes sense. I don't think I'll roll that into this release, but might consider it for a future one. Depends on if I get the time to implement configurable drag-to-snap areas.

**Comment by kamzata** [comment]
> @kamzata Gotcha. That makes sense. I don't think I'll roll that into this release, but might consider it for a future one. Depends on if I get the time to implement configurable drag-to-snap areas.

Anyway, 4 windows increase the productivity... it would be interesting it could be set up also as 50%-25%-25% or 25%-50%-25% or 25%-25%-50%. It could be also interesting the Eighth, thus splitting vertically like for the Sixth. All these things should be "snappable" of course. It would be great.

**Comment by rxhanson** [comment]
@kamzata Right before working on these additional window sizes for Rectangle, I added a "make your own shortcut" feature to [Hookshot](https://hookshot.app) (my paid alternative). You can find it in the wrench & ruler tab. This is functionality I was looking to add into Rectangle at first, but it just doesn't fit with the simplicity of Rectangle. That's why I decided to add the additional window sizes to Rectangle - I still want it to be useful to users with large monitors but without complexity.

I'm going to write a detailed post explaining the "make your own shortcut feature" in Hookshot, but the gist is this: For a keyboard shortcut, you can specify any window size and snap location, and provide additional window sizes and locations for repeated executions of that shortcut. The UI could use a little work, since it's my first cut, but the functionality is there.
Further customizations of the "Hookshot" functionality and the drag-to-snap areas are on my roadmap, but I don't have a timeline for those. 

As mentioned, I have no plans of adding this level of customization into Rectangle, and don't really see adding any other window sizes in the foreseeable future.

**Comment by xsanda** [comment]
> Mind giving the beta a test? https://github.com/rxhanson/Rectangle/releases/tag/0.33

This works very well for me! My only feedback would be possibly to have the option to minimise sections of the menu, as it’s getting rather crowded: it might be worth hiding the lesser-used ones in submenus. As I mainly use the menu to check what the shortcut is or to use options I don’t often use, I wouldn’t mind them being hidden behind an extra click/hover.

**Comment by rxhanson** [comment]
@xsanda Thanks for the feedback! I've been trying to figure out an elegant way to do this, and I think you're spot on with hiding the lesser-used ones in submenus. Automatically hiding lesser used ones might be a little difficult to get right (as in, maybe I don't want top left quarter to go away, even though I don't commonly use it, etc). I might just let the ones under the prefs window fold (thirds, fourths, move, and sixths) to be in submenus at first cut.

**Comment by xsanda** [comment]
> @xsanda Thanks for the feedback! I've been trying to figure out an elegant way to do this, and I think you're spot on with hiding the lesser-used ones in submenus. Automatically hiding lesser used ones might be a little difficult to get right (as in, maybe I don't want top left quarter to go away, even though I don't commonly use it, etc). I might just let the ones under the prefs window fold (thirds, fourths, move, and sixths) to be in submenus at first cut.

That seems sensible to me

**Comment by StErMi** [comment]
> @kamzata Right before working on these additional window sizes for Rectangle, I added a "make your own shortcut" feature to [Hookshot](https://hookshot.app) (my paid alternative). You can find it in the wrench & ruler tab. This is functionality I was looking to add into Rectangle at first, but it just doesn't fit with the simplicity of Rectangle. That's why I decided to add the additional window sizes to Rectangle - I still want it to be useful to users with large monitors but without complexity.
> 
> I'm going to write a detailed post explaining the "make your own shortcut feature" in Hookshot, but the gist is this: For a keyboard shortcut, you can specify any window size and snap location, and provide additional window sizes and locations for repeated executions of that shortcut. The UI could use a little work, since it's my first cut, but the functionality is there.
> Further customizations of the "Hookshot" functionality and the drag-to-snap areas are on my roadmap, but I don't have a timeline for those.
> 
> As mentioned, I have no plans of adding this level of customization into Rectangle, and don't really see adding any other window sizes in the foreseeable future.

so Hookshot is like Rectangle but with more features? or am I getting it wrong? I mean if I purchase it I can do the same thing but with more feature? I was looking for something that specifically support 49" 32:9 monitors like mine ;)

**Comment by StErMi** [comment]
I've installed the beta version and all the commands like 

- first third / last third 
- first fourth / last fourth

works the first time but if you click them another wine they put the same windows on the next spot, I don't know if this is the correct behavior but it seems to do that.

It would be cool to also have, as requested in the thread something that allows me to choose a predefined layout like 25-50-25 or a combination of it or 25-25-25-25 and it will create some "hard" zones where windows that get dropped over there will respect that aspect ratio.

Also, I saw that "reactive points" (I don't know how you call them) that will come up when I move the window in that point only works at the bottom of the screen and not in the top (it's a little problem for me because I have the 49" above the MacBook Pro so I would prefer to have those active areas in the top of the screen).

Another feature but I don't know if can be implemented in macOS would be that the MacBook menu that will show up in the header of the screen based on the active windows can be replicated above each section based on which is the active app in that specific section of the screen.

**Comment by rxhanson** [comment]
@StErMi 

> so Hookshot is like Rectangle but with more features? or am I getting it wrong? I mean if I purchase it I can do the same thing but with more feature? I was looking for something that specifically support 49" 32:9 monitors like mine ;)

Yes. Hookshot is Rectangle but with more features. It's built on top of Rectangle, and I cherry pick all of Rectangle's new features into Hookshot. There are a couple things that Hookshot doesn't have, like localization. Almost all of the terminal commands for Rectangle are also available for Hookshot as well (just replace 'Rectangle' with 'Hookshot'). I built Hookshot for myself and liked it enough to turn it into its own product. Personally, I prefer it over using touch gestures for window management.

> works the first time but if you click them another wine they put the same windows on the next spot, I don't know if this is the correct behavior but it seems to do that.

This is intentional. You can disable the repeated actions with a command [listed on the readme](https://github.com/rxhanson/Rectangle#halves-to-thirds-repeated-execution-of-half-and-quarter-actions).

> Also, I saw that "reactive points" (I don't know how you call them) that will come up when I move the window in that point only works at the bottom of the screen and not in the top (it's a little problem for me because I have the 49" above the MacBook Pro so I would prefer to have those active areas in the top of the screen).

I think you're referring to the drag-to-snap areas? The top of the screen is for maximize. I might add in the ability to customize these in a future release of Hookshot.

>Another feature but I don't know if can be implemented in macOS would be that the MacBook menu that will show up in the header of the screen based on the active windows can be replicated above each section based on which is the active app in that specific section of the screen.

I'm not quite sure I understand. Perhaps a picture might help?

**Comment by StErMi** [comment]
> > Another feature but I don't know if can be implemented in macOS would be that the MacBook menu that will show up in the header of the screen based on the active windows can be replicated above each section based on which is the active app in that specific section of the screen.
> 
> I'm not quite sure I understand. Perhaps a picture might help?

I have divided the screen into 3 but the top bar only "works" for the selected windows. it would be cool (with monitors this big) to have it replicated like I would have three real monitors.

![image](https://user-images.githubusercontent.com/146166/95008135-5e3cf600-0617-11eb-9e51-05619351b64b.png)

Another feature that would be cool is to have "hard" zones that emulate the physical screen. So if you drag your windows in those zones they can only expand at max into the zone (like if you have real monitors but without borders and bezels)

**Comment by rxhanson** [comment]
> I have divided the screen into 3 but the top bar only "works" for the selected windows. it would be cool (with monitors this big) to have it replicated like I would have three real monitors.

Ah ok, I get what you mean. Technically, I think that should be possible to implement. That's an interesting idea, but I don't know if I'd add it into any of my existing apps. 

> Another feature that would be cool is to have "hard" zones that emulate the physical screen. So if you drag your windows in those zones they can only expand at max into the zone (like if you have real monitors but without borders and bezels)

I have some ideas that I was looking to implement in Hookshot that are somewhat similar to this. I'll keep this in mind when I go down that path.

For any future thoughts, feel free to create a new issue so that it's easy for people to find common feature ideas on the repo (that way we don't deviate too far on this issue as well).

**Comment by xsanda** [comment]
@StErMi What you're referring to would probably have to be handled by the video card or monitor, as it would require macOS to believe the display was multiple monitors. The only example of anyone ever implementing this properly would be an [LG monitor with multiple physical HDMI ports](http://radnex.us/LG_43UN700-B_Multi-Input-Display), so entirely performed in hardware.

See also: [A StackOverflow post asking this question](https://apple.stackexchange.com/a/397184)

**Comment by rxhanson** [comment]
This is now out of beta in v0.34. Please open a new issue if there are any issues with it.

**Comment by StErMi** [comment]
@rxhanson would you like me to create another issue for the feature request?

**Comment by rxhanson** [comment]
@StErMi I don't think it's necessary, but you can if you want. I'm aware of it and will see if it makes sense to add it into Hookshot when I add in other features (no guarantee). I don't really see adding this functionality into Rectangle since that would add complexity to the app that I don't really want to add in.

**Comment by seanpianka** [comment]
Is there any support for splitting into eighths? I also have a 32:9 monitor that I want to cycle through halved-fourths on the top and bottom, but it seems sixths is the maximum support (or split into fourths and manually resize to half of the vertical)

**Comment by xsanda** [comment]
> Is there any support for splitting into eighths? I also have a 32:9 monitor that I want to cycle through halved-fourths on the top and bottom, but it seems sixths is the maximum support (or split into fourths and manually resize to half of the vertical)

@seanpianka The most user-friendly way I can imagine implementing this, without adding 8-10 new keyboard shortcuts, would be to have keyboard shortcuts for maintaining the existing width, but setting the height. For example, I use ⌃⌥1-4 for the four quarters, which could be followed by ⌃⌥⌘↑ or ⌃⌥⌘↓ to divide the height. Technically you’d only need one of these, as a second press of it would get you to the other position.

Come to think of it, if the thirds/quarters commands work the other way around on vertical monitors, it may be worth this command working the other way around too on vertical monitors.

While I’ve never really worked with Swift, I’d be more than willing to give this a go in a PR if @rxhanson is interested, as it’s something I've felt could be useful too.

**Comment by rxhanson** [comment]
@xsanda I'm actually not 100% sure what you're looking to add, but it sounds like something along the lines of actions for "dividing size in direction". I can see accepting a pull request with something like this, but I think I'll need a better understanding of what you're thinking. Would a user be able to configure the divide actions? Do the divide actions only divide once?

**Comment by xsanda** [comment]
@rxhanson the command would preserve the current width and horizontal position, but set the height to 50% of the screen, and the vertical position against the top edge. Press it once more, and it would move to the bottom edge. Press it once more and it would restore to the original size.

I'm not sure what it would be called, maybe Move to Top Half? There could likewise be a Move to Bottom Half for inverting the cycle.

**Comment by rxhanson** [comment]
@xrr Got it, thanks for the diagram. I think the simplest approach from a user perspective would be to just modify the move left/right/up/down actions with a configurable percent of the screen that the window dimension would take depending on the direction moved. Note that it's on my todo to change the default behavior of the move left/right/up/down actions so that they don't center (the logic is there, it's currently toggled via terminal command). 

This way, we don't have to introduce more window action shortcuts, and we could even make it so that the repeated executions can mostly match what you have there. The configurable percent of the screen for the dimension would just be a terminal config at first, and I'm open to having things added to the UI only if there's enough demand. Let me know if this doesn't make sense and I can clarify. 

If you were bent on having new window actions, I'd still consider a PR but not with any UI changes at first, meaning everything would be configurable via terminal and could have some hardcoded values for the shortcuts. This is a simpler route in some ways than modifying the move left/right/up/down actions.

If you are ready to give something a go, just let me know the direction you want to take it and I can give you some entry points and model behavior in the code.

Just food for thought, my first reaction was the idea of implementing new actions that would snap a window to the edge in a direction and half the dimension (% of screen) each time it was executed. So you'd end up with something where "twice" in your diagram would be a quarter screen height bordering the top, etc. I saw another window manager do something like that once, although I can't remember which one (there's too many out there).

**Comment by xsanda** [comment]
@rxhanson overloading the move up/down/left/right commands is a good idea, it probably makes more sense than adding yet more commands to the menu. I’ll have a look at that tomorrow.

Re the point of infinite halving, that could work (⌃⌥4, ⌃⌥⌘↑, ⌃⌥⌘↓ would for example put you in the 4th column, third row), but personally I think I’d prefer to copy the behaviour of single-screen left/right (_When unchecked, repeated half actions cycle ½, ⅔, and ⅓_). It does mean you can split by thirds rather than fourths, but I imagine that’s probably preferable in the shorter direction of a monitor.

I’ve put in a draft PR with my work from today (#283), feel free to look over it now just to see how I’ve tackled it, even if a fair amount of it is going to be rewritten. One particular point to note is that this command performs resizing in one dimension but not the other, which means that, in order to prevent the window shrinking by (`2 × gap`) each time the command is pressed, I had to increase the width by `2 × gap` in the Calculation. Come to think of it, it would probably be preferable instead to update `WindowAction#gapsApplicable`* to work on a per-dimension basis.

---

To summarise, my current thoughts are to add a command-line switch `resizeOnDirectionalMove` or similar, which, when enabled, would perform the triple-state (½, ⅔, and ⅓) cycle on directional moves.

---

\* I noticed that `gapsApplicable` omitted `.centerHalf` (and `gapSharedEdge` is missing quite a few cases), which I’m happy to fix in this PR or another intermediate one.
